File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/SEORoute.vb
Public Class SEORoute
    Inherits Route

    Public Sub New(ByVal url As String, ByVal routeHandler As IRouteHandler)
        MyBase.New(url, routeHandler)
    End Sub

    Public Sub New(ByVal url As String, ByVal defaults As RouteValueDictionary, ByVal routeHandler As IRouteHandler)
        MyBase.New(url, defaults, routeHandler)
    End Sub

    Public Sub New(ByVal url As String, ByVal defaults As RouteValueDictionary, ByVal constraints As RouteValueDictionary, ByVal routeHandler As IRouteHandler)
        MyBase.New(url, defaults, constraints, routeHandler)
    End Sub

    Public Sub New(ByVal url As String, ByVal defaults As RouteValueDictionary, ByVal constraints As RouteValueDictionary, ByVal dataTokens As RouteValueDictionary, ByVal routeHandler As IRouteHandler)
        MyBase.New(url, defaults, constraints, dataTokens, routeHandler)
    End Sub

    Public Overrides Function GetVirtualPath(ByVal requestContext As System.Web.Routing.RequestContext, ByVal values As System.Web.Routing.RouteValueDictionary) As System.Web.Routing.VirtualPathData
        Dim path As VirtualPathData = MyBase.GetVirtualPath(requestContext, values)

        If Not path Is Nothing Then
            path.VirtualPath = path.VirtualPath.AppendTrailingSlash()
        End If

        Return path
    End Function
End Class

File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/SEORoute.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Routing;

namespace CoolMvcBlog.Utils
{
    public class SEORoute : Route
    {
        public SEORoute(string url, IRouteHandler routeHandler)
			: base(url, routeHandler) { }
		public SEORoute(string url, RouteValueDictionary defaults, IRouteHandler routeHandler)
			: base(url, defaults, routeHandler) { }
		public SEORoute(string url, RouteValueDictionary defaults, RouteValueDictionary constraints, IRouteHandler routeHandler)
			: base(url, defaults, constraints, routeHandler) { }
        public SEORoute(string url, RouteValueDictionary defaults, RouteValueDictionary constraints, RouteValueDictionary dataTokens, IRouteHandler routeHandler)
			: base(url, defaults, constraints, dataTokens, routeHandler) { }

		public override VirtualPathData GetVirtualPath(
            RequestContext requestContext, RouteValueDictionary values)
		{
			VirtualPathData path = base.GetVirtualPath(requestContext, values);
            
            if (path != null)
            {
                path.VirtualPath = path.VirtualPath.AppendTrailingSlash();
            }

			return path;
		}
    }
}